File: src/games/sudoku/gameplay/analysis/techniqueModules/coloring.ts
import { sudokuDigits } from '../../../types';
import { cellPeers } from '../bitmask';
import { buildCandidateEliminationMove, collectHousesFromIndexes, getHouseDigitMatches } from '../techniqueHelpers';
import type { SudokuTechniqueDispatcher } from '../techniqueModuleTypes';

function addEdge(graph: Map<number, Set<number>>, left: number, right: number): void {
  graph.set(left, new Set([...(graph.get(left) ?? []), right]));
  graph.set(right, new Set([...(graph.get(right) ?? []), left]));
}

export const coloringTechnique: SudokuTechniqueDispatcher = {
  technique: 'coloring',
  tier: 'expert',
  findMove(state) {
    const houses = [
      ...Array.from({ length: 9 }, (_, index) => Array.from({ length: 9 }, (_, col) => (index * 9) + col)),
      ...Array.from({ length: 9 }, (_, index) => Array.from({ length: 9 }, (_, row) => (row * 9) + index)),
      ...Array.from({ length: 9 }, (_, box) => {
        const rowStart = Math.floor(box / 3) * 3;
        const colStart = (box % 3) * 3;
        return Array.from({ length: 9 }, (_, offset) => (
          ((rowStart + Math.floor(offset / 3)) * 9) + colStart + (offset % 3)
        ));
      }),
    ];

    for (const digit of sudokuDigits) {
      const graph = new Map<number, Set<number>>();
      houses.forEach((houseCells) => {
        const matches = getHouseDigitMatches(state, houseCells, digit);
        if (matches.length === 2) {
          addEdge(graph, matches[0], matches[1]);
        }
      });

      const seen = new Set<number>();
      for (const start of graph.keys()) {
        if (seen.has(start)) {
          continue;
        }

        const colors = new Map<number, 0 | 1>([[start, 0]]);
        const queue = [start];
        const component: number[] = [];

        while (queue.length > 0) {
          const current = queue.shift();
          if (typeof current === 'undefined' || seen.has(current)) {
            continue;
          }
          seen.add(current);
          component.push(current);

          const currentColor = colors.get(current) ?? 0;
          (graph.get(current) ?? []).forEach((neighbor) => {
            if (!colors.has(neighbor)) {
              colors.set(neighbor, currentColor === 0 ? 1 : 0);
            }
            if (!seen.has(neighbor)) {
              queue.push(neighbor);
            }
          });
        }

        for (const color of [0, 1] as const) {
          const colorCells = component.filter((index) => colors.get(index) === color);
          const hasConflict = colorCells.some((left, leftIndex) => (
            colorCells.slice(leftIndex + 1).some((right) => cellPeers[left].includes(right))
          ));
          if (!hasConflict) {
            continue;
          }

          const move = buildCandidateEliminationMove({
            technique: 'coloring',
            eliminations: colorCells.map((index) => ({ index, digit })),
            evidenceCells: component,
            houses: collectHousesFromIndexes(component),
          });
          if (move) {
            return move;
          }
        }

        const colorZero = component.filter((index) => colors.get(index) === 0);
        const colorOne = component.filter((index) => colors.get(index) === 1);
        const move = buildCandidateEliminationMove({
          technique: 'coloring',
          eliminations: state.board
            .map((value, index) => ({ value, index }))
            .filter(({ value, index }) => value === 0 && !component.includes(index))
            .filter(({ index }) => getHouseDigitMatches(state, [index], digit).length === 1)
            .filter(({ index }) => (
              colorZero.some((colorIndex) => cellPeers[index].includes(colorIndex))
              && colorOne.some((colorIndex) => cellPeers[index].includes(colorIndex))
            ))
            .map(({ index }) => ({ index, digit })),
          evidenceCells: component,
          houses: collectHousesFromIndexes(component),
        });
        if (move) {
          return move;
        }
      }
    }

    return null;
  },
};

File: engine/src/main/kotlin/com/quietgrid/engine/sudoku/techniques/Coloring.kt
package com.quietgrid.engine.sudoku.techniques

import com.quietgrid.engine.core.Difficulty
import com.quietgrid.engine.sudoku.*
import java.util.ArrayDeque

object ColoringTechnique : SudokuTechniqueDispatcher {
    override val technique = SudokuTechnique.COLORING
    override val tier = Difficulty.EXPERT

    private fun addEdge(graph: MutableMap<Int, MutableSet<Int>>, left: Int, right: Int) {
        graph.getOrPut(left) { mutableSetOf() }.add(right)
        graph.getOrPut(right) { mutableSetOf() }.add(left)
    }

    override fun findMove(state: SudokuBitmaskState): SudokuCanonicalMove? {
        var best: SudokuCanonicalMove? = null
        var bestComplexity = Int.MAX_VALUE

        for (digit in 1..9) {
            val graph = mutableMapOf<Int, MutableSet<Int>>()
            (rowCellIndexes + columnCellIndexes + boxCellIndexes).forEach { houseCells ->
                val matches = getHouseDigitMatches(state, houseCells, digit)
                if (matches.size == 2) addEdge(graph, matches[0], matches[1])
            }

            val seen = mutableSetOf<Int>()
            for (start in graph.keys) {
                if (start in seen) continue

                val colors = mutableMapOf(start to 0)
                val queue = ArrayDeque<Int>()
                queue.add(start)
                val component = mutableListOf<Int>()

                while (queue.isNotEmpty()) {
                    val current = queue.poll()
                    if (current in seen) continue
                    seen.add(current)
                    component.add(current)
                    val currentColor = colors[current] ?: 0
                    (graph[current] ?: emptySet()).forEach { neighbor ->
                        if (neighbor !in colors) colors[neighbor] = if (currentColor == 0) 1 else 0
                        if (neighbor !in seen) queue.add(neighbor)
                    }
                }

                val complexity = component.size
                if (complexity >= bestComplexity) continue

                for (color in listOf(0, 1)) {
                    val colorCells = component.filter { colors[it] == color }
                    val hasConflict = colorCells.indices.any { leftIndex ->
                        colorCells.drop(leftIndex + 1).any { right -> cellPeers[colorCells[leftIndex]].contains(right) }
                    }
                    if (!hasConflict) continue

                    val move = buildCandidateEliminationMove(
                        SudokuTechnique.COLORING, colorCells.map { it to digit }, component, collectHousesFromIndexes(component), complexity,
                    )
                    if (move != null) { best = move; bestComplexity = complexity; break }
                }
                if (complexity >= bestComplexity) continue

                val colorZero = component.filter { colors[it] == 0 }
                val colorOne = component.filter { colors[it] == 1 }
                val eliminations = state.board.indices
                    .filter { state.board[it] == 0 && it !in component }
                    .filter { getHouseDigitMatches(state, listOf(it), digit).size == 1 }
                    .filter { index -> colorZero.any { cellPeers[index].contains(it) } && colorOne.any { cellPeers[index].contains(it) } }
                    .map { it to digit }
                val move = buildCandidateEliminationMove(SudokuTechnique.COLORING, eliminations, component, collectHousesFromIndexes(component), complexity)
                if (move != null) { best = move; bestComplexity = complexity }
            }
        }
        return best
    }
}
